daisy/workflow.go

```diff
@@ -841,8 +841,20 @@ func readWorkflow(file string, w *Workflow) DError {
 		return newErr("failed to get absolute path of workflow file", err)
 	}
 
-	if err := json.Unmarshal(data, &w); err != nil {
-		return newErr("failed to unmarshal workflow file", JSONError(file, data, err))
+	// Set input vars and run var substitution before unmarshal.
+	if varMap != nil {
```

Comment:
I think this can be simplified using bytes.ReplaceAll, strings.ReplaceAll _may_ cost a bit more but i didn't look into the implementation too deeply (use whichever you think is more readable).

```
for k, v := range varMap {
  wfJSON = bytes.RepalceAll(wfJSON, []byte(fmt.Sprintf("${%s}", k)), []byte(v))
}
```

Reply:
Here we used strings.Replacer other than RepalceAll. Implementation of strings.Replacer is based on bytes array, which write only one pass. It's a O(n) alogorithm.

bytes.RepalceAll in a for-loop is a O(n*m) algorithm, assuming m = loop times.

Reply:
ack, thought Replacer ran in order